File: backend/app/engines/candidates.py
"""CandidateGenerator: which ticketed destinations C are worth a paid shop for A→B?

The naive plan shops A→every airport. We rank C by what our own database has
learned and shop only the top few. OpenFlights B→C spokes are a cold-start hint,
not a schedule; observed A→B→C tickets outweigh them as soon as we have any.

candidateScore =
      0.30 × historical_connection_probability   (A→C tickets that route via B)
    + 0.25 × historical_savings_probability      (…and undercut the honest A→B)
    + 0.20 × expected_savings                    (median saving % of A→B, clipped)
    + 0.10 × freshness                           (how recently that happened)
    + 0.10 × carrier_hub_probability             (OpenFlights published B→C)
    + 0.05 × provider_success_rate               (our suppliers actually answer A→C)
"""

from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field
from datetime import datetime, timezone

WEIGHTS = {
    "connection": 0.30,
    "savings_probability": 0.25,
    "expected_savings": 0.20,
    "freshness": 0.10,
    "hub": 0.10,
    "provider": 0.05,
}
DEFAULT_MIN_SCORE = 0.35
COLD_START_PROBES = 3
DEAD_AFTER_CHECKS = 5  # checked this many times, never routed via B → stop paying for it
FRESHNESS_HALF_LIFE_DAYS = 7.0


@dataclass(slots=True)
class RouteStat:
    """One hidden_city_route_stats row, provider-neutral."""

    origin: str
    intended: str
    ticketed: str
    observations: int = 0
    successful_connections: int = 0
    cheaper_than_direct_count: int = 0
    median_saving: float = 0.0
    average_saving_percent: float = 0.0
    last_success_at: datetime | None = None
    last_cheaper_at: datetime | None = None


@dataclass(slots=True)
class Candidate:
    code: str
    score: float
    source: str
    parts: dict[str, float] = field(default_factory=dict)
    observations: int = 0
    successful_connections: int = 0
    cheaper_than_direct_count: int = 0
    median_saving: float = 0.0
    selected: bool = False


def _smooth(successes: int, trials: int) -> float:
    """Laplace-style rate so one lucky hit is not a 100% probability."""
    if trials <= 0:
        return 0.0
    return (successes + 0.5) / (trials + 1.0)


def freshness(last: datetime | None, now: datetime) -> float:
    if last is None:
        return 0.0
    age_days = max((now - last).total_seconds(), 0.0) / 86400.0
    return 0.5 ** (age_days / FRESHNESS_HALF_LIFE_DAYS)


def score_candidate(
    stat: RouteStat | None,
    hub_probability: float,
    provider_rate: float,
    now: datetime,
) -> tuple[float, dict[str, float]]:
    if stat is None:
        parts = {
            "connection": 0.0,
            "savings_probability": 0.0,
            "expected_savings": 0.0,
            "freshness": 0.0,
            "hub": hub_probability,
            "provider": provider_rate,
        }
    else:
        parts = {
            "connection": _smooth(stat.successful_connections, stat.observations),
            "savings_probability": _smooth(stat.cheaper_than_direct_count, stat.observations),
            "expected_savings": max(0.0, min(stat.average_saving_percent / 100.0, 1.0)),
            "freshness": freshness(stat.last_cheaper_at or stat.last_success_at, now),
            "hub": hub_probability,
            "provider": provider_rate,
        }
    score = sum(WEIGHTS[k] * v for k, v in parts.items())
    return round(score, 4), {k: round(v, 4) for k, v in parts.items()}


def is_dead(stat: RouteStat | None) -> bool:
    return bool(stat and stat.observations >= DEAD_AFTER_CHECKS and stat.successful_connections == 0)


def hub_probabilities(intended: list[str] | set[str], route_map: dict[str, set[str]] | None = None) -> dict[str, float]:
    """1.0 when OpenFlights published a historical B→C nonstop; else 0. Not a schedule."""
    out: dict[str, float] = defaultdict(float)
    for b in intended:
        for c in (route_map or {}).get(b.upper(), set()):
            out[c.upper()] = 1.0
    return dict(out)


def rank_candidates(
    origin: str,
    intended: list[str] | set[str],
    stats: dict[str, RouteStat],
    pool: dict[str, str],
    hub_prob: dict[str, float],
    provider_rate: dict[str, float],
    now: datetime | None = None,
) -> list[Candidate]:
    """Score every C in `pool` (code → where it came from) plus every C we hold stats for."""
    now = now or datetime.now(timezone.utc)
    origin_u = origin.upper()
    excluded = {origin_u, *(b.upper() for b in intended)}
    codes: dict[str, str] = {}
    for code, source in pool.items():
        c = code.upper()
        if len(c) == 3 and c not in excluded:
            codes.setdefault(c, source)
    for c in stats:
        if len(c) == 3 and c not in excluded:
            codes.setdefault(c.upper(), "stats")
    out: list[Candidate] = []
    for code, source in codes.items():
        stat = stats.get(code)
        if is_dead(stat):
            continue
        score, parts = score_candidate(
            stat, hub_prob.get(code, 0.0), provider_rate.get(code, 0.5), now
        )
        out.append(
            Candidate(
                code=code,
                score=score,
                source="stats" if stat and stat.observations else source,
                parts=parts,
                observations=stat.observations if stat else 0,
                successful_connections=stat.successful_connections if stat else 0,
                cheaper_than_direct_count=stat.cheaper_than_direct_count if stat else 0,
                median_saving=stat.median_saving if stat else 0.0,
            )
        )
    out.sort(key=lambda c: (-c.score, -c.observations, c.code))
    return out


def select_candidates(
    ranked: list[Candidate],
    budget: int,
    min_score: float = DEFAULT_MIN_SCORE,
    cold_start: int = COLD_START_PROBES,
    exclude: set[str] | None = None,
) -> list[Candidate]:
    """Top-N above the threshold; on a cold start still probe a few hubs so we learn."""
    if budget <= 0:
        return []
    skip = {c.upper() for c in (exclude or set())}
    usable = [c for c in ranked if c.code not in skip]
    chosen = [c for c in usable if c.score >= min_score][:budget]
    if len(chosen) < min(cold_start, budget):
        picked = {c.code for c in chosen}
        for c in usable:
            if len(chosen) >= min(cold_start, budget):
                break
            if c.code not in picked:
                chosen.append(c)
                picked.add(c.code)
    for c in ranked:
        c.selected = c in chosen
    return chosen


def expected_value(p_useful: float, p_purchase: float, commission: float) -> float:
    """EV of one more paid request. Compare with settings.search_cost_usd."""
    return round(max(p_useful, 0.0) * max(p_purchase, 0.0) * max(commission, 0.0), 4)


# ── legacy in-process graph (kept for the current-process debug view) ──────────


@dataclass
class HiddenCityCandidateEdge:
    origin: str
    connection: str
    ticketed_destination: str
    observation_count: int = 0
    last_observed_at: str = ""
    best_observed_price: float | None = None


_GRAPH: dict[tuple[str, str, str], HiddenCityCandidateEdge] = {}


def record_offer(origin: str, connection: str, ticketed: str, price: float | None) -> None:
    key = (origin.upper(), connection.upper(), ticketed.upper())
    now = datetime.now(timezone.utc).isoformat(timespec="seconds")
    edge = _GRAPH.get(key)
    if edge is None:
        _GRAPH[key] = HiddenCityCandidateEdge(
            origin=key[0],
            connection=key[1],
            ticketed_destination=key[2],
            observation_count=1,
            last_observed_at=now,
            best_observed_price=price,
        )
    else:
        edge.observation_count += 1
        edge.last_observed_at = now
        if price is not None and (edge.best_observed_price is None or price < edge.best_observed_price):
            edge.best_observed_price = price


def graph_snapshot() -> list[dict]:
    return [
        {
            "origin": e.origin,
            "connection": e.connection,
            "ticketed_destination": e.ticketed_destination,
            "observation_count": e.observation_count,
            "last_observed_at": e.last_observed_at,
            "best_observed_price": e.best_observed_price,
        }
        for e in sorted(_GRAPH.values(), key=lambda x: (-x.observation_count, x.origin))
    ]


def candidates_beyond(
    origin: str,
    intended: list[str] | set[str],
    limit: int,
    extra: list[str] | None = None,
) -> list[str]:
    """Cold-start order with no database: observed this process, then caller extras."""
    intended_u = {code.upper() for code in intended}
    origin_u = origin.upper()
    seen = {origin_u, *intended_u}
    ranked: list[str] = []

    def push(codes: list[str]) -> None:
        for code in codes:
            code = code.upper()
            if len(code) == 3 and code not in seen:
                seen.add(code)
                ranked.append(code)

    observed = [
        e.ticketed_destination
        for e in sorted(_GRAPH.values(), key=lambda x: -x.observation_count)
        if e.origin == origin_u and e.connection in intended_u
    ]
    push(observed)
    push(list(extra or []))
    return ranked[:limit]


def plan_expansion(
    origin: str,
    intended: list[str] | set[str],
    limit: int,
    extra: list[str] | None = None,
    reused: list | None = None,
    already_hidden: bool = False,
) -> tuple[list[str], list[str]]:
    """Destinations still worth shopping, plus ticketed C values already indexed through B."""
    from app.engines.index import ticketed_dests_through

    dests = {code.upper() for code in intended}
    covered = ticketed_dests_through(reused or [], dests)
    skipped = sorted(covered)
    if already_hidden or covered or limit <= 0:
        return [], skipped
    ranked = candidates_beyond(origin, intended, limit, extra=extra)
    return ranked[:limit], skipped

File: backend/src/engines/candidates.rs
use chrono::{DateTime, Utc};
use std::collections::{HashMap, HashSet};
use std::sync::{LazyLock, Mutex};

use crate::engines::index::ticketed_dests_through;
use crate::models::Offer;

pub static WEIGHTS: LazyLock<HashMap<&'static str, f64>> = LazyLock::new(|| {
    HashMap::from([
        ("connection", 0.30),
        ("savings_probability", 0.25),
        ("expected_savings", 0.20),
        ("freshness", 0.10),
        ("hub", 0.10),
        ("provider", 0.05),
    ])
});

pub const DEFAULT_MIN_SCORE: f64 = 0.35;
/// Default cold-start fill in unit tests. Live search passes FAST_CANDIDATES (floor 5).
pub const COLD_START_PROBES: usize = 3;
pub const DEAD_AFTER_CHECKS: i32 = 5;
pub const FRESHNESS_HALF_LIFE_DAYS: f64 = 7.0;
/// Frozen OpenFlights (~2014–2017) is a structural prior, not a timetable.
pub const OPENFLIGHTS_HUB: f64 = 0.35;
/// B→C seen inside a priced ticket, or a hidden-city stat row.
pub const LEARNED_HUB: f64 = 1.0;

pub fn source_priority(source: &str) -> i32 {
    match source {
        "stats" => 4,
        "index" => 3,
        "edges" => 2,
        "openflights" | "hub" => 0,
        _ => 1,
    }
}

pub fn is_learned_source(source: &str) -> bool {
    matches!(source, "stats" | "index" | "edges")
}

/// Keep the stronger planner source. OpenFlights never overwrites learned rows.
pub fn assign_source(pool: &mut HashMap<String, String>, code: String, source: &str) {
    let incoming = source_priority(source);
    let current = pool.get(&code).map(|s| source_priority(s)).unwrap_or(-1);
    if incoming > current {
        pool.insert(code, source.to_string());
    }
}

fn is_learned_candidate(c: &Candidate) -> bool {
    c.observations > 0 || is_learned_source(&c.source)
}

#[derive(Debug, Clone)]
pub struct RouteStat {
    pub origin: String,
    pub intended: String,
    pub ticketed: String,
    pub observations: i32,
    pub successful_connections: i32,
    pub cheaper_than_direct_count: i32,
    pub median_saving: f64,
    pub average_saving_percent: f64,
    pub last_success_at: Option<DateTime<Utc>>,
    pub last_cheaper_at: Option<DateTime<Utc>>,
}

impl RouteStat {
    pub fn new(origin: &str, intended: &str, ticketed: &str) -> Self {
        Self {
            origin: origin.to_string(),
            intended: intended.to_string(),
            ticketed: ticketed.to_string(),
            observations: 0,
            successful_connections: 0,
            cheaper_than_direct_count: 0,
            median_saving: 0.0,
            average_saving_percent: 0.0,
            last_success_at: None,
            last_cheaper_at: None,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Candidate {
    pub code: String,
    pub score: f64,
    pub source: String,
    pub parts: HashMap<String, f64>,
    pub observations: i32,
    pub successful_connections: i32,
    pub cheaper_than_direct_count: i32,
    pub median_saving: f64,
    pub selected: bool,
}

impl Candidate {
    pub fn new(code: &str, score: f64, source: &str) -> Self {
        Self {
            code: code.to_string(),
            score,
            source: source.to_string(),
            parts: HashMap::new(),
            observations: 0,
            successful_connections: 0,
            cheaper_than_direct_count: 0,
            median_saving: 0.0,
            selected: false,
        }
    }
}

fn smooth(successes: i32, trials: i32) -> f64 {
    if trials <= 0 {
        0.0
    } else {
        (successes as f64 + 0.5) / (trials as f64 + 1.0)
    }
}

pub fn freshness(last: Option<DateTime<Utc>>, now: DateTime<Utc>) -> f64 {
    let last = match last {
        Some(v) => v,
        None => return 0.0,
    };
    let age_days = (now - last).num_seconds().max(0) as f64 / 86400.0;
    0.5_f64.powf(age_days / FRESHNESS_HALF_LIFE_DAYS)
}

pub fn score_candidate(
    stat: Option<&RouteStat>,
    hub_probability: f64,
    provider_rate: f64,
    now: DateTime<Utc>,
) -> (f64, HashMap<String, f64>) {
    let parts = if let Some(stat) = stat {
        HashMap::from([
            (
                "connection".to_string(),
                smooth(stat.successful_connections, stat.observations),
            ),
            (
                "savings_probability".to_string(),
                smooth(stat.cheaper_than_direct_count, stat.observations),
            ),
            (
                "expected_savings".to_string(),
                (stat.average_saving_percent / 100.0).clamp(0.0, 1.0),
            ),
            (
                "freshness".to_string(),
                freshness(stat.last_cheaper_at.or(stat.last_success_at), now),
            ),
            ("hub".to_string(), hub_probability),
            ("provider".to_string(), provider_rate),
        ])
    } else {
        HashMap::from([
            ("connection".to_string(), 0.0),
            ("savings_probability".to_string(), 0.0),
            ("expected_savings".to_string(), 0.0),
            ("freshness".to_string(), 0.0),
            ("hub".to_string(), hub_probability),
            ("provider".to_string(), provider_rate),
        ])
    };
    let score: f64 = WEIGHTS
        .iter()
        .map(|(k, w)| w * parts.get(*k).copied().unwrap_or(0.0))
        .sum();
    let rounded: HashMap<String, f64> = parts
        .into_iter()
        .map(|(k, v)| (k, (v * 10000.0).round() / 10000.0))
        .collect();
    ((score * 10000.0).round() / 10000.0, rounded)
}

pub fn is_dead(stat: Option<&RouteStat>) -> bool {
    stat.map(|s| s.observations >= DEAD_AFTER_CHECKS && s.successful_connections == 0)
        .unwrap_or(false)
}

pub fn hub_probabilities(
    intended: &HashSet<String>,
    openflights: &HashMap<String, HashSet<String>>,
    learned: &HashSet<String>,
) -> HashMap<String, f64> {
    let mut out: HashMap<String, f64> = HashMap::new();
    for b in intended {
        if let Some(cs) = openflights.get(&b.to_uppercase()) {
            for c in cs {
                out.insert(c.to_uppercase(), OPENFLIGHTS_HUB);
            }
        }
    }
    for c in learned {
        let c = c.to_uppercase();
        if c.len() == 3 {
            out.insert(c, LEARNED_HUB);
        }
    }
    out
}

pub fn rank_candidates(
    origin: &str,
    intended: &HashSet<String>,
    stats: &HashMap<String, RouteStat>,
    pool: &HashMap<String, String>,
    hub_prob: &HashMap<String, f64>,
    provider_rate: &HashMap<String, f64>,
    now: Option<DateTime<Utc>>,
) -> Vec<Candidate> {
    let now = now.unwrap_or_else(Utc::now);
    let origin_u = origin.to_uppercase();
    let mut excluded: HashSet<String> = intended.iter().map(|b| b.to_uppercase()).collect();
    excluded.insert(origin_u);
    let mut codes: HashMap<String, String> = HashMap::new();
    for (code, source) in pool {
        let c = code.to_uppercase();
        if c.len() == 3 && !excluded.contains(&c) {
            assign_source(&mut codes, c, source);
        }
    }
    for c in stats.keys() {
        let c = c.to_uppercase();
        if c.len() == 3 && !excluded.contains(&c) {
            codes.insert(c, "stats".into());
        }
    }
    let mut out = Vec::new();
    for (code, source) in codes {
        let stat = stats.get(&code);
        if is_dead(stat) {
            continue;
        }
        let (score, parts) = score_candidate(
            stat,
            hub_prob.get(&code).copied().unwrap_or(0.0),
            provider_rate.get(&code).copied().unwrap_or(0.5),
            now,
        );
        out.push(Candidate {
            code: code.clone(),
            score,
            source: if stat.map(|s| s.observations > 0).unwrap_or(false) {
                "stats".into()
            } else {
                source
            },
            parts,
            observations: stat.map(|s| s.observations).unwrap_or(0),
            successful_connections: stat.map(|s| s.successful_connections).unwrap_or(0),
            cheaper_than_direct_count: stat.map(|s| s.cheaper_than_direct_count).unwrap_or(0),
            median_saving: stat.map(|s| s.median_saving).unwrap_or(0.0),
            selected: false,
        });
    }
    out.sort_by(|a, b| {
        b.score
            .partial_cmp(&a.score)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then_with(|| source_priority(&b.source).cmp(&source_priority(&a.source)))
            .then_with(|| b.observations.cmp(&a.observations))
            .then_with(|| a.code.cmp(&b.code))
    });
    out
}

pub fn select_candidates(
    ranked: &mut [Candidate],
    budget: usize,
    min_score: f64,
    cold_start: usize,
    exclude: &HashSet<String>,
) -> Vec<Candidate> {
    if budget == 0 {
        return vec![];
    }
    let skip: HashSet<String> = exclude.iter().map(|c| c.to_uppercase()).collect();
    let usable: Vec<usize> = ranked
        .iter()
        .enumerate()
        .filter(|(_, c)| !skip.contains(&c.code))
        .map(|(i, _)| i)
        .collect();
    let mut chosen_idx: Vec<usize> = usable
        .iter()
        .copied()
        .filter(|i| ranked[*i].score >= min_score)
        .take(budget)
        .collect();
    let has_learned = usable.iter().any(|i| is_learned_candidate(&ranked[*i]));
    let fill_to = cold_start.min(budget);
    if chosen_idx.len() < fill_to {
        let mut picked: HashSet<String> =
            chosen_idx.iter().map(|i| ranked[*i].code.clone()).collect();
        for i in &usable {
            if chosen_idx.len() >= fill_to {
                break;
            }
            if picked.contains(&ranked[*i].code) {
                continue;
            }
            if has_learned && !is_learned_candidate(&ranked[*i]) {
                continue;
            }
            picked.insert(ranked[*i].code.clone());
            chosen_idx.push(*i);
        }
    }
    for c in ranked.iter_mut() {
        c.selected = false;
    }
    for i in &chosen_idx {
        ranked[*i].selected = true;
    }
    chosen_idx.into_iter().map(|i| ranked[i].clone()).collect()
}

pub fn expected_value(p_useful: f64, p_purchase: f64, commission: f64) -> f64 {
    ((p_useful.max(0.0) * p_purchase.max(0.0) * commission.max(0.0)) * 10000.0).round() / 10000.0
}

#[derive(Debug, Clone)]
pub struct HiddenCityCandidateEdge {
    pub origin: String,
    pub connection: String,
    pub ticketed_destination: String,
    pub observation_count: i32,
    pub last_observed_at: String,
    pub best_observed_price: Option<f64>,
}

static GRAPH: LazyLock<Mutex<HashMap<(String, String, String), HiddenCityCandidateEdge>>> =
    LazyLock::new(|| Mutex::new(HashMap::new()));

pub fn record_offer(origin: &str, connection: &str, ticketed: &str, price: Option<f64>) {
    let key = (
        origin.to_uppercase(),
        connection.to_uppercase(),
        ticketed.to_uppercase(),
    );
    let now = Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Secs, true);
    let mut g = crate::mutex_lock(&GRAPH);
    if let Some(edge) = g.get_mut(&key) {
        edge.observation_count += 1;
        edge.last_observed_at = now;
        if let Some(p) = price {
            if edge.best_observed_price.map(|b| p < b).unwrap_or(true) {
                edge.best_observed_price = Some(p);
            }
        }
    } else {
        g.insert(
            key.clone(),
            HiddenCityCandidateEdge {
                origin: key.0,
                connection: key.1,
                ticketed_destination: key.2,
                observation_count: 1,
                last_observed_at: now,
                best_observed_price: price,
            },
        );
    }
}

pub fn graph_snapshot() -> Vec<serde_json::Value> {
    let mut edges: Vec<_> = crate::mutex_lock(&GRAPH).values().cloned().collect();
    edges.sort_by(|a, b| {
        b.observation_count
            .cmp(&a.observation_count)
            .then_with(|| a.origin.cmp(&b.origin))
    });
    edges
        .into_iter()
        .map(|e| {
            serde_json::json!({
                "origin": e.origin,
                "connection": e.connection,
                "ticketed_destination": e.ticketed_destination,
                "observation_count": e.observation_count,
                "last_observed_at": e.last_observed_at,
                "best_observed_price": e.best_observed_price,
            })
        })
        .collect()
}

pub fn candidates_beyond(
    origin: &str,
    intended: &HashSet<String>,
    limit: usize,
    extra: Option<&[String]>,
) -> Vec<String> {
    let intended_u: HashSet<String> = intended.iter().map(|c| c.to_uppercase()).collect();
    let origin_u = origin.to_uppercase();
    let mut seen = intended_u.clone();
    seen.insert(origin_u.clone());
    let mut ranked = Vec::new();
    let mut push = |codes: Vec<String>| {
        for code in codes {
            let code = code.to_uppercase();
            if code.len() == 3 && seen.insert(code.clone()) {
                ranked.push(code);
            }
        }
    };
    let mut observed: Vec<(i32, String)> = crate::mutex_lock(&GRAPH)
        .values()
        .filter(|e| e.origin == origin_u && intended_u.contains(&e.connection))
        .map(|e| (e.observation_count, e.ticketed_destination.clone()))
        .collect();
    observed.sort_by(|a, b| b.0.cmp(&a.0));
    push(observed.into_iter().map(|(_, c)| c).collect());
    if let Some(extra) = extra {
        push(extra.to_vec());
    }
    ranked.truncate(limit);
    ranked
}

pub fn plan_expansion(
    origin: &str,
    intended: &HashSet<String>,
    limit: usize,
    extra: Option<&[String]>,
    reused: &[Offer],
    already_hidden: bool,
) -> (Vec<String>, Vec<String>) {
    let dests: HashSet<String> = intended.iter().map(|c| c.to_uppercase()).collect();
    let covered = ticketed_dests_through(reused, &dests);
    let mut skipped: Vec<String> = covered.iter().cloned().collect();
    skipped.sort();
    if already_hidden || limit == 0 {
        return (vec![], skipped);
    }
    let dests: Vec<String> = candidates_beyond(origin, intended, limit + covered.len(), extra)
        .into_iter()
        .filter(|c| !covered.contains(c))
        .take(limit)
        .collect();
    (dests, skipped)
}
